### PR TITLE
*: use patch instead of component

### LINKED ISSUE #3887: Extract a new component for common dependencies
There are some crates that are dependented by multiple components, eg, both `tikv`, `test_coprocessor`, `test_storage` and `test_raftstore` depends on `kvproto`, as a result, when we upgrade or test another branch of `kvproto`, we have to change theirs `Cargo.toml`s. It would be convenient to upgrade or test new branch if we add a new component which exports `kvproto` and any other components depend on it.

### PR DESCRIPTION
**Pull request by BusyJay**
Patch is the documented way to replace a dependency with a patched one.
So it's recommended to be used instead of component. Besides, using the
same name for component package can confuse compiler sometimes and
result in compilation error.

For example, if you want to update kvproto, you can add following lines to
Cargo.toml in the root directory:

```
[patch.'https://github.com/pingcap/kvproto.git']
kvproto = { git = "my url", branch = "my branch" }
```

Close #3887.

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -890,21 +890,14 @@ dependencies = [
 [[package]]
 name = "kvproto"
 version = "0.0.1"
-source = "git+https://github.com/pingcap/kvproto.git#d4fe6b3363790236f9df4e04dce5b4674921a134"
+source = "git+https://github.com/busyjay/kvproto.git?branch=test-patch#48951a223cdd549c03c59eec6030e59bd62db863"
```
Comment (zhangjinpeng1987) [inline Cargo.lock]: Please revert this change.

Comment (siddontang) [reply Cargo.lock]: +1

**Comment by breeswish** [comment]
Looks nice! Shall we add RocksDB to the patch list as well?
